src/agent/src/watcher.rs

```diff
@@ -255,38 +278,59 @@ impl SandboxStorages {
-                // If an error was observed, we will stop treating this Storage as being watchable, and
-                // instead clean up the target-mount files on the tmpfs and bind mount the source_mount_point
-                // to target_mount_point.
-                error!(logger, "error observed when watching: {:?}", e);
-                entry.watch = false;
-
-                // Remove destination contents, but not the directory itself, since this is
-                // assumed to be bind-mounted into a container. If source/mount is a file, no need to cleanup
-                if entry.target_mount_point.as_path().is_dir() {
-                    for dir_entry in std::fs::read_dir(entry.target_mount_point.as_path())? {
-                        let dir_entry = dir_entry?;
-                        let path = dir_entry.path();
-                        if dir_entry.file_type()?.is_dir() {
-                            tokio::fs::remove_dir_all(path).await?;
-                        } else {
-                            tokio::fs::remove_file(path).await?;
+                match e.downcast_ref::<WatcherError>() {
+                    Some(WatcherError::MountTooLarge { .. })
+                    | Some(WatcherError::MountTooManyFiles { .. }) => {
+                        //
+                        // If the mount we were watching is too large (bytes), or contains too many unique files,
+                        // we no longer want to watch. Instead, we'll attempt to create a bind mount and mark this storage
+                        // as non-watchable. if there's an error in creating bind mount, we'll continue watching.
+                        //
+                        // Ensure the target mount point exists:
+                        if !entry.target_mount_point.as_path().exists() {
+                            if entry.source_mount_point.as_path().is_dir() {
+                                fs::create_dir_all(entry.target_mount_point.as_path())
+                                    .await
+                                    .with_context(|| {
+                                        format!(
+                                            "create dir for bindmount {:?}",
+                                            entry.target_mount_point.as_path()
+                                        )
+                                    })?;
+                            } else {
+                                fs::File::create(entry.target_mount_point.as_path())
+                                    .await
+                                    .with_context(|| {
+                                        format!(
+                                            "create file {:?}",
+                                            entry.target_mount_point.as_path()
+                                        )
+                                    })?;
+                            }
+                        }
+
+                        match BareMount::new(
```

Comment:
where do we unmount the mountpoint? Looks like it is missing a cleanup method?

Reply:
Yes, it seemed the cleanup fn didn't take care those mounts's umount.

Reply:
Good call; thanks ya'll.  Pushed additional commit to add unit test for this, as well as do the cleanup. PTAL!